x/stakeibc/keeper/icqcallbacks_delegator_shares.go

```diff
@@ -39,49 +41,78 @@ func DelegatorSharesCallback(k Keeper, ctx sdk.Context, args []byte, query icqty
 	}
 	k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Delegation, "Query response - Delegator: %s, Validator: %s, Shares: %v",
-		queriedDelgation.DelegatorAddress, queriedDelgation.ValidatorAddress, queriedDelgation.Shares))
+		queriedDelegation.DelegatorAddress, queriedDelegation.ValidatorAddress, queriedDelegation.Shares))
 
-	// Ensure ICQ can be issued now, else fail the callback
-	withinBufferWindow, err := k.IsWithinBufferWindow(ctx)
-	if err != nil {
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "unable to determine if ICQ callback is inside buffer window, err: %s", err.Error())
-	}
-	if !withinBufferWindow {
-		return errorsmod.Wrapf(types.ErrOutsideIcqWindow, "callback is outside ICQ window")
+	// Unmarshal the callback data containing the previous delegation to the validator (from the time the query was submitted)
+	var callbackData types.DelegatorSharesQueryCallback
+	if err := proto.Unmarshal(query.CallbackData, &callbackData); err != nil {
+		return errorsmod.Wrapf(err, "unable to unmarshal delegator shares callback data")
 	}
 
 	// Grab the validator object from the hostZone using the address returned from the query
-	validator, valIndex, found := GetValidatorFromAddress(hostZone.Validators, queriedDelgation.ValidatorAddress)
+	_, valIndex, found := GetValidatorFromAddress(hostZone.Validators, queriedDelegation.ValidatorAddress)
 	if !found {
-		return errorsmod.Wrapf(types.ErrValidatorNotFound, "no registered validator for address (%s)", queriedDelgation.ValidatorAddress)
+		return errorsmod.Wrapf(types.ErrValidatorNotFound, "no registered validator for address (%s)", queriedDelegation.ValidatorAddress)
 	}
 
-	// Get the validator's internal exchange rate, aborting if it hasn't been updated this epoch
-	strideEpochTracker, found := k.GetEpochTracker(ctx, epochtypes.STRIDE_EPOCH)
-	if !found {
-		return errorsmod.Wrapf(sdkerrors.ErrNotFound, "unable to get epoch tracker for epoch (%s)", epochtypes.STRIDE_EPOCH)
+	// Confirm the validator was slashed by looking at the number of tokens associated with the delegation
+	validatorWasSlashed, delegatedTokens, err := k.CheckForSlash(ctx, hostZone, valIndex, queriedDelegation)
+	if err != nil {
+		return err
+	}
+	// If the validator was not slashed, exit now
+	if !validatorWasSlashed {
+		return nil
 	}
-	if validator.InternalExchangeRate.EpochNumber != strideEpochTracker.EpochNumber {
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest,
-			"validator (%s) internal exchange rate has not been updated this epoch (epoch #%d)", validator.Address, strideEpochTracker.EpochNumber)
+
+	// Check if the ICQ overlapped a delegation or undelegation ICA that would have modfied the number of delegated tokens
+	icaOverlappedIcq, err := k.CheckDelegationChangeWhileQueryInFlight(ctx, hostZone, valIndex, query, callbackData)
+	if err != nil {
+		return err
+	}
+	// If the ICA/ICQ overlapped, a new query will be submitted, exit now
```

Comment:
Does IBC transfers to the delegation account not count as a ICA action.  If someone gives us an LSM token and we send it back over to the hub immediately that doesn't affect this?  It is only actions like a delegate, begin unbond, or send account to account on the hub which would matter for resetting this.... then yes not as worried since no external person can force those actions

Reply:
yes that's right IBC transfers are different from ICAs....

but I just realized the detokenizations will change our delegation 🤦

Reply:
OK I'm going to leave this one open because it seems like a real issue we still have to think through

Reply:
This is going to require going back to the drawing board so I’m going to resolve this for the sake of merging this PR

Reply:
this wound up not being as a big of a deal as originally anticipated - resolved by #759